include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+    return !tl.empty() && tl[tl.ndim()-1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsImage(const TensorLayout &tl) {
+    return NumSpatialDims(tl) >= 2;
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes a video
+ */
+struct VideoLayoutInfo : ImageLayoutInfo {
+  /** @brief Returns the index of the dimension referring to frames */
+  DALI_HOST_DEV
+  static int FrameDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'F');
+  }
+  DALI_HOST_DEV
+  static bool IsChannelFirst(const TensorLayout &tl) {
+    return tl[FrameDimIndex(tl)+1] == 'C';
+  }
+  DALI_HOST_DEV
+  static bool IsSequence(const TensorLayout &tl) {
+    return tl.contains('F');
+  }
+  DALI_HOST_DEV
+  static bool IsVideo(const TensorLayout &tl) {
+    return IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static bool IsStillImage(const TensorLayout &tl) {
+    return !IsSequence(tl) && IsImage(tl);
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetFrameLayout(const TensorLayout &tl) {
+    return tl.skip('F');
+  }
+  DALI_HOST_DEV
+  static TensorLayout GetSequenceLayout(const TensorLayout &tl) {
+    if (tl.contains('F'))
+      return tl;
+    if (tl[0] == 'N')
+      return "NF" + tl.sub(1);
+    else
+      return "F" + tl;
+  }
+};
+
+/** @brief Calculates mapping of dimensions from given output to input.
+ *
+ * Array element corresponding to given output dimensions is the index of that
+ * dimension in the input layout.
+ *
+ * Example:
+ * ```
+ * input_layout = "NHWC"
+ * output_layout = "NCHW"
+ * result = { 0, 3, 1, 2 }
+ * ```
+ * Explanation:
+ *   - 'N' appears at index 0 in both input_layout and output_layout.
+ *   - 'C' appears at index 3 in input
+ *   - 'H' appears at index 1 in input
+ *   - 'W' appears at index 2 in input
+ *
+ * If there are repetitions, the repeated names are listed in increasing order, e.g.
+ * ```
+ *  input_layout = "aaabb"
+ *  output_layout = "baaba"
+ *  result = { 3, 0, 1, 4, 2 }
+ * ```
+ */
+template <int Dims>
+inline std::array<int, Dims> permuted_dims(const TensorLayout &in_layout,
+                                           const TensorLayout &out_layout) {
+  std::array<int, Dims> perm_dims;
+  for (int d = 0; d < Dims; d++) {
+    perm_dims[d] = d;
+  }
+
+  if (in_layout.empty() || out_layout.empty() || in_layout == out_layout)
+    return perm_dims;
+
+  DALI_ENFORCE(in_layout.ndim() == Dims && out_layout.ndim() == Dims,
+    "Unexpected number of dimensions in layout description");
+
+  const char *a = out_layout.data();
+  TensorLayout b = in_layout;
+
+  int min_j = 0;
+  for (int i = 0; i < Dims; i++) {
+    char c = a[i];
+    int j;
+    while (b[min_j] == 0 && min_j < Dims)
+      min_j++;
+    for (j = min_j; j < Dims; j++) {
+      if (b[j] == c)
+        break;
+    }
+    if (j == Dims)
+      DALI_FAIL("\"" + out_layout.str() + "\" is not a permutation of \"" + in_layout.str() + "\"");
```

Comment:
```DALI_ENFORCE(j != Dims, ...)```

Reply:
`j == Dims` or `j != Dims` is meaningless to the user - it would only distract them from the real problem (the fact that one layout is not a permutation of the other).